pjsip/src/pjsip/sip_multipart.c

```diff
@@ -78,17 +78,48 @@ static int multipart_print_body(struct pjsip_msg_body *msg_body,
 	p += m_data->boundary.slen;
 	*p++ = 13; *p++ = 10;
 
+	/* Determine whether we're going to add an explicit Content-Type
+	 * to the bodypart
+	 */
+	pj_bool_t insert_ct = (part->body && part->body->content_type.type.slen);
+
 	/* Print optional headers */
 	hdr = part->hdr.next;
 	while (hdr != &part->hdr) {
-	    int printed = pjsip_hdr_print_on((pjsip_hdr*)hdr, p,
+		/* Don't print header if we're about to override it */
+		if (!insert_ct || (hdr->type != PJSIP_H_CONTENT_TYPE)) {
+			int printed = pjsip_hdr_print_on((pjsip_hdr*)hdr, p,
 	                                     SIZE_LEFT()-2);
-	    if (printed < 0)
-		return -1;
-	    p += printed;
-	    *p++ = '\r';
-	    *p++ = '\n';
-	    hdr = hdr->next;
+			if (printed < 0)
+				return -1;
+			p += printed;
+			*p++ = '\r';
+			*p++ = '\n';
+		}
+		hdr = hdr->next;
+	}
+
+	/* Automatically add Content-Type header, only
+	 * if content_type is set in the message body.
+	 */
+	if (insert_ct) {
+		pj_str_t ctype_hdr = { "Content-Type: ", 14};
```

Comment:
I suspect you've just resurrected the original code, but it's quite sad that it's basically a straight copy from https://github.com/Metaswitch/pjsip-upstream/blob/8588ad804ddb759b1f8b5225121dd6b485336562/pjsip/src/pjsip/sip_msg.c#L508-L524

In fact, quite a lot of this code is just straight copied from there. Sigh.

Reply:
I feel uncomfortable about trying to refactor this code to avoid duplication (would rather change as little of this code as possible).  I'd feel easier leaving as is - let me know if you're particularly keen for me to fix it up